File: iteration2/Advisor.java
package iteration2;

import java.util.ArrayList;
import java.util.List;

/**
 * Represents an adviser who can log in, reach request of the student, approve and reject request of the student
 * Inherits from the Person class and implements the IDisplayMenu interface.
 */
public class Advisor extends Lecturer implements IDisplayMenu {

    private List<Student> studentsAdvised;
    private List<Registration> requestList;
    private int requestNumber;
    private Notification notification;


    /**
     * Creates a new advisor object with the given parameters
     *
     * @param ID       the advisor's ID
     * @param name     the advisor's name
     * @param surname  the advisor's surname
     * @param userName the advisor's userName
     * @param password the advisor's password
     */
    public Advisor(int ID, String name, String surname, String userName, String password) {
        super(ID, name, surname, userName, password);
        studentsAdvised = new ArrayList<>();
        requestList = new ArrayList<>();
    }

    /**
     * The method prints the first five request of students
     * Then prompts to advisor choose one of them to evaluate
     * Then prints the detailed request of selected student
     * And calls replyRequest method
     */
    public void printRequests() {
        CourseRegistrationSystem controller = new CourseRegistrationSystem();
        ConsoleColours.paintBlueMenu();
        System.out.println("Request Approval Menu:");
        System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨\n");
        System.out.println("             Back -> 0");
        System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
        if (!requestList.isEmpty()) {
            System.out.println("Request(s) Listed Below:");
            System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");

            ConsoleColours.paintPurpleMenu();
            for (int i = 0; i < requestList.size(); i++) {
                System.out.println((i + 1) + ". " + requestList.get(i).getStudent().getName()
                        + " " + requestList.get(i).getStudent().getSurname());
                System.out.println("``````````````````````````````````````````````");
            }

            ConsoleColours.paintGreenMenu();
            System.out.println("Enter the request number you want to evaluate: ");
            requestNumber = controller.getInput();

            ConsoleColours.paintBlueMenu();
            if (requestNumber <= requestList.size() && requestNumber >= 1) {
                Registration request = requestList.get(requestNumber - 1);
                System.out.println(request.getStudent().getName() + " "
                        + request.getStudent().getSurname() + " wants to take these courses:");
                ConsoleColours.paintPurpleMenu();
                for (Course course : request.getCourses()) {
                    if (!(request.getStudent().getTranscript().getCourseGradeMap().get(course) != null
                            && request.getStudent().getTranscript().getCourseGradeMap().get(course).getLast() == null))
                        System.out.println(course.getCourseCode() + " " + course.getCourseName());
                }
                ConsoleColours.paintBlueMenu();
                System.out.println(request.getStudent().getName() + " "
                        + request.getStudent().getSurname() + " wants to drop these courses:");
                ConsoleColours.paintPurpleMenu();
                for (Course course : request.getCourses()) {
                    if ((request.getStudent().getTranscript().getCourseGradeMap().get(course) != null
                            && request.getStudent().getTranscript().getCourseGradeMap().get(course).getLast() == null))
                        System.out.println(course.getCourseCode() + " " + course.getCourseName());
                }
                replyRequests();
            } else if (requestNumber > requestList.size() || requestNumber < 0) {
                ConsoleColours.paintRedMenu();
                System.out.println("Invalid choice Please select again!");
                printRequests();
            }
        } else {
            ConsoleColours.paintYellowMenu();
            System.out.println("There is not a Waiting Request!");
            ConsoleColours.resetColour();
            System.out.println();
        }
    }

    /**
     * The method prompts to adviser select an option for request
     * Calls the appropriate methods from registration class
     */
    public void replyRequests() {
        CourseRegistrationSystem controller = new CourseRegistrationSystem();

        ConsoleColours.paintBlueMenu();
        System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
        System.out.println("               Back -> 0");
        System.out.println("    Approve Request -> 1");
        System.out.println("     Reject Request -> 2");
        System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
        ConsoleColours.paintGreenMenu();
        System.out.println("Please select an option:");

        int choice = controller.getInput();
        switch (choice) {
            case 1:
                requestList.get(requestNumber - 1).approveRequest();
                requestList.remove(requestNumber - 1);
                printRequests();
                break;
            case 2:
                requestList.get(requestNumber - 1).rejectRequest();
                requestList.remove(requestNumber - 1);
                printRequests();
                break;
            case 0:
                printRequests();
                break;
            default:
                ConsoleColours.paintRedMenu();
                System.out.println("Invalid choice. Please select again!");
                replyRequests();
                break;
        }
    }

    /**
     * Prints the advisor menu
     * Prompts to advisor select an action
     */
    @Override
    public void printMenu(String menuType) {
        ConsoleColours.paintBlueMenu();
        System.out.println("Welcome " + getName() + " " + getSurname() + "!");
        System.out.println("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~");

        ConsoleColours.paintYellowMenu();
        String message = "You have " + requestList.size() + " request(s).";
        System.out.println(message);
        ConsoleColours.paintBlueMenu();

        System.out.println("Please select from the following options:");
        System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
        System.out.println("                                Exit -> 0");
        System.out.println("                       List requests -> 1");
        System.out.println("                             Log out -> 2");
        System.out.println("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨");
        ConsoleColours.paintGreenMenu();
        System.out.print("Enter your choice: \n");
    }

    /**
     * The method to check whether the userName and password match
     */
    @Override
    public boolean login(String userName, String password) {
        return getUserName().equals(userName) && getPassword().equals(password);
    }

    public List<Student> getStudentsAdvised() {
        return studentsAdvised;
    }

    public List<Registration> getRequests() {
        return requestList;
    }

    public List<Registration> getRequestList() {
        return requestList;
    }

    public void setNotification(Notification notification) {
        this.notification = notification;
    }
}

File: iteration3/Advisor.py
import logging

from iteration3.IDisplayMenu import IDisplayMenu
from iteration3.Lecturer import Lecturer
from iteration3.ConsoleColours import ConsoleColours


class Advisor(Lecturer, IDisplayMenu):

    def __init__(self, ID, name, surname, userName, password):
        super().__init__(ID, name, surname, userName, password)
        from iteration3.CourseRegistrationSystem import CourseRegistrationSystem
        self.__system = CourseRegistrationSystem()
        self.__studentsAdvised = []
        self.__requestList = []
        self.__requestNumber = None
        self.__notification = None

    def printRequests(self):
        ConsoleColours.paintBlueMenu()
        print("Request Approval Menu:")
        print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨\n")
        print("\nBack -> 0\n")
        print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
        if len(self.__requestList) != 0:
            print("Request(s) Listed Below:")
            print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
            ConsoleColours.paintPurpleMenu()
            for i in range(len(self.__requestList)):
                print(
                    f"{i + 1}. {self.__requestList[i].getStudent().getName()} {self.__requestList[i].getStudent().getSurname()}")
                print("``````````````````````````````````````````````")

                ConsoleColours.paintGreenMenu()
                print("Enter the request number you want to evaluate: ")

            requestNumber = self.__system.getInput()
            ConsoleColours.paintGreenMenu()
            if 1 <= requestNumber <= len(self.__requestList):
                self.__requestNumber = requestNumber
                request = self.__requestList[requestNumber - 1]
                print(
                    f"{request.getStudent().getName()} {request.getStudent().getSurname()} wants to take these courses:")
                ConsoleColours.paintPurpleMenu()
                self.checkAndPrintOverlap(request)
                ConsoleColours.paintBlueMenu()
                print()
                ConsoleColours.paintRedMenu()
                print(
                    f"{request.getStudent().getName()} {request.getStudent().getSurname()} wants to drop these courses:")
                ConsoleColours.paintPurpleMenu()
                for course in request.getCourses():
                    if request.getStudent().getTranscript().getCourseGradeMap().get(course) is not None \
                            and request.getStudent().getTranscript().getCourseGradeMap().get(course)[-1] is None:
                        print(f"{course.getCourseCode()} {course.getCourseName()}")
                self.replyRequests()
            elif requestNumber > len(self.__requestList) or requestNumber < 0:
                ConsoleColours.paintRedMenu()
                logging.warning("Advisor " + self.getID() + " entered invalid input.")
                print("Invalid choice! Please select again.")
                self.printRequests()
        else:
            ConsoleColours.paintYellowMenu()
            print("There is not a waiting request!")
            print("You are redirecting to the Advisor Main Menu...")
            ConsoleColours.resetColour()
            print()

    def checkAndPrintOverlap(self, request):
        sorted_courses = []
        nested_list = []
        for course in request.getCourses():
            sorted_courses.append(course.getCourseCode())
            sorted_courses.append(course.getCourseName())
            sorted_courses.append(course.getDay())
            sorted_courses.append(course.getHour())
            nested_list.append(sorted_courses)
            sorted_courses = []
            if not (request.getStudent().getTranscript().getCourseGradeMap().get(course) is not None
                    and request.getStudent().getTranscript().getCourseGradeMap().get(course)[-1] is None):
                print(
                    f"{course.getCourseCode()} {course.getCourseName()} - {course.getDay()} - {course.getHour()}.00")
                print()
        lanet = []
        for i in range(len(nested_list) - 1):
            conflict = 0
            if i in lanet:
                continue
            controlInput = 0
            for j in range(i + 1, len(nested_list)):
                if j in lanet:
                    continue
                first_list = nested_list[i]
                second_list = nested_list[j]

                es_Day = first_list[2]
                es_Hour = first_list[3]

                g_day = second_list[2]
                g_Hour = second_list[-1]
                ConsoleColours.paintRedMenu()
                if (es_Day == g_day) and (es_Hour == g_Hour) and controlInput == 0:
                    print(
                        f"{first_list[0]} - {first_list[1]} / {second_list[0]} - {second_list[1]} /", end='')
                    controlInput = 1
                    conflict = 1
                    lanet.append(j)
                    continue
                if (es_Day == g_day) and (
                        es_Hour == g_Hour) and controlInput == 1:
                    print(
                        f"{second_list[0]} - {second_list[1]} /",
                        end='')
                    lanet.append(j)
                    conflict = 1
            if conflict == 1:
                print(
                    f"------->{nested_list[i][2]} - {nested_list[i][3]}.00 -----> CLASS OVERLAP!!!")

    def replyRequests(self):
        ConsoleColours.paintBlueMenu()
        print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
        print("               Back -> 0")
        print("    Approve Request -> 1")
        print("     Reject Request -> 2")
        print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
        ConsoleColours.paintGreenMenu()
        print("Please select an option:")

        choice = self.__system.getInput()
        if choice == 1:
            self.__requestList[self.__requestNumber - 1].approveRequest()
            logging.info(
                f"Advisor {self.getID()} approved request of student {self.__requestList[self.__requestNumber - 1].getStudent().getID()}")
            self.__requestList.pop(self.__requestNumber - 1)
            self.printRequests()
        elif choice == 2:
            self.__requestList[self.__requestNumber - 1].rejectRequest()
            logging.info(
                f"Advisor {self.getID()} rejected request of student {self.__requestList[self.__requestNumber - 1].getStudent().getID()}")
            self.__requestList.pop(self.__requestNumber - 1)
            self.printRequests()
        elif choice == 0:
            self.printRequests()
        else:
            ConsoleColours.paintRedMenu()
            logging.warning(f"Advisor {self.getID()} entered invalid input.")
            print("Invalid choice. Please select again!")
            self.replyRequests()

    def printMenu(self, menuType):
        ConsoleColours.paintBlueMenu()
        print(f"Welcome {self.getName()} {self.getSurname()}!")
        print("~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~")

        ConsoleColours.paintYellowMenu()
        message = f"You have {len(self.__requestList)} request(s)."
        print(message)
        print()
        ConsoleColours.paintBlueMenu()

        print("Please select from the following options:")
        print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
        print("                                Exit -> 0")
        print("                       List requests -> 1")
        print("                             Log out -> 2")
        print("¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨¨")
        ConsoleColours.paintGreenMenu()
        print("Enter your choice: ")

    def login(self, userName, password):
        return self.getUserName() == userName and self.getPassword() == password

    def getStudentsAdvised(self):
        return self.__studentsAdvised

    def getRequestList(self):
        return self.__requestList

    def setNotification(self, notification):
        self.__notification = notification
